pkg/xds/server/callbacks/dataplane_callbacks.go

```diff
@@ -54,22 +52,40 @@ type dpStream struct {
 var _ util_xds.Callbacks = &xdsCallbacks{}
 
 func (d *xdsCallbacks) OnStreamClosed(streamID core_xds.StreamID) {
-	var lastStreamDpKey *core_model.ResourceKey
+	var streamDpKey *core_model.ResourceKey
+	d.RLock()
+	dpStream := d.dpStreams[streamID]
+	streamDpKey = dpStream.dp
+	d.RUnlock()
+
+	if streamDpKey != nil {
+		// execute callback after lock is freed, so heavy callback implementation won't block every callback for every DPP.
+		// proxy disconnect may take some time, we don't want to block the callback for every DPP here
+		disconnectDone := make(chan struct{})
+		d.callbacks.OnProxyDisconnected(dpStream.ctx, streamID, *streamDpKey, func() {
+			go func() { disconnectDone <- struct{}{} }()
+		})
+		go func() {
+			<-disconnectDone
+			d.cleanupDpStream(streamID)
+		}()
```

Comment:
It seems like before calling `d.cleanupDpStream(streamID)` we may call `OnStreamOpen` first, then we gonna hit concurrent problem.